src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -858,249 +1078,112 @@ protected SCMRevision retrieve(SCMHead head, TaskListener listener) throws IOExc
+                    try {
+                        GitHub github = Connector.connect(apiUri, credentials);
+                        try {
+                            checkApiUrlValidity(github);
+                            Connector.checkApiRateLimit(listener, github);
+                            ghRepository = github.getRepository(fullName);
+                            LOGGER.log(Level.INFO, "Got remote pull requests from {0}", fullName);
+                            int n = 0;
+                            for (GHPullRequest pr: ghRepository.queryPullRequests().state(GHIssueState.OPEN).list()) {
+                                pullRequestSourceMap.put(pr.getNumber(), new PullRequestSource(
+                                        pr.getHead().getRepository().getOwnerName(),
+                                        pr.getHead().getRepository().getName(),
+                                        pr.getHead().getRef()));
+                                n++;
+                                if (n % 30  == 0) { // default page size is 30
+                                    Connector.checkApiRateLimit(listener, github);
+                                }
+                            }
+                        } finally {
+                            Connector.release(github);
+                        }
+                    } catch (IOException | InterruptedException e) {
+                        LOGGER.log(Level.WARNING,
+                                "Could not get all pull requests from " + fullName + ", there may be rebuilds", e);
+                    }
+                }
             }
-            // hack!
-            return repositoryUrl.toExternalForm().replace(repoOwner+"/"+repository, owner+"/"+repo);
-        }
-        if (StringUtils.isBlank(apiUri)) {
-            return "https://github.com/"+ owner+"/"+repo;
-        }
-        if (StringUtils.endsWith(StringUtils.removeEnd(apiUri, "/"), "/api/v3")) {
-            return StringUtils.removeEnd(StringUtils.removeEnd(apiUri, "/"), "api/v3") + owner + "/" + repo;
+            return pullRequestSourceMap.get(number);
         }
-        return null;
     }
 
     /**
-     * Similar to {@link PreBuildMerge}, but we cannot use that unmodified: we need to specify the exact base branch hash.
-     * It is possible to just ask Git to check out {@code refs/pull/123/merge}, but this has two problems:
-     * GitHub’s implementation is not all that reliable (for example JENKINS-33237, and time-delayed snapshots);
-     * and it is subject to a race condition between the {@code baseHash} we think we are merging with and a possibly newer one that was just pushed.
-     * Performing the merge ourselves is simple enough and ensures that we are building exactly what the {@link PullRequestSCMRevision} represented.
+     * Retained to migrate legacy configuration.
+     * @deprecated use {@link MergeWithGitSCMExtension}.
      */
+    @Restricted(DoNotUse.class)
+    @RestrictedSince("3.4.0")
```

Comment:
ahhh intellij thank you for annotating overriding methods with the same annotations